src/openbsd.h

```diff
@@ -5,19 +5,24 @@
 #include <cstddef>
 #include <cstring>
 #include <ctime>
+#include <sys/time.h>
+#include <sys/types.h>
+#include <unistd.h>
 
 inline
 void arc4random_buf(void *buf, size_t nbytes)
 {
-
     for( size_t n = 0; n < nbytes; ++ n)
         ((char*)(buf))[n] = rand() %256;
 }
 
 inline
 void arc4random_init(void)
 {
-    srand( (unsigned int) time(NULL));
+  struct timeval tv;
+  gettimeofday(&tv, 0);
+  // this is not very good, but we lack a portable non-blocking API
+  srand( ((unsigned int) tv.tv_usec) ^ (unsigned int)getpid());
```

Comment:
I appreciate the problem, and I also value the simplicity. But: on the operating system lending this file its name, there's a non-blocking `/dev/urandom`. That's not POSIX, but quite the same on Linux, [FreeBSD](https://man.freebsd.org/cgi/man.cgi?query=urandom&sektion=4&manpath=FreeBSD+4.6-RELEASE), and as far as I know also MacOS, and its various mutations.

Is there a strong argument for doing a `gettimeofday` syscall and a `getpid` syscall to supply low-quality randomness, rather than `open`, `read`, and `close` to get a good seed?

Reply:
the key point is "as far as we know". Also this code might be used in places that can't see /dev. The seed is not a secret it just needs to vary a bit. I think this library has chosen to be "something that just works", and then this choice is good enough.

Reply:
I see, though I'd like to point out that one of the common places where this kind of time-based seeding fails short is on booting of embedded devices without RTC. E.g., router turns on, takes (down to few milliseconds) exactly the same time to load initrd, boot kernel, launch admin interface, which launches the "first user setup" routine, which hence always ends up with the same time-based seed. I don't know how PIDs are randomized on such a system.

Now, you'll be right to exclaim "but a few milliseconds is still way much variation in µs!", and you'd be very right. But, fun awaits at least under Linux with userland time functions on platforms that don't have a hardware high-res timer register; [in many cases, time gets systick-quantized](https://unix.stackexchange.com/questions/759880/why-is-filesystem-time-always-some-msecs-behind-system-time-in-linux/759894#759894), even on x86_64.

Hence my caveat: I 100% would merge this, it's indubitably an improvement, but for the cases where a non-varying seed is especially problematic, namely, when someone with the same device can reconstruct the same seed, I'm not sure it's a solution :)

Reply:
Hi guys,

thank you discussing and supporting this project.
This author did some investigations around that:

https://codingnest.com/generating-random-numbers-using-c-standard-library-the-problems/

It looks like there is not simple way to implement a platform independent solution.
/dev/urandom and gettimeofday would prevent us using the code within MS Windows.